Add Komiic source

SummonHIM opened the pull request:
Checklist:

- [x] Updated `extVersionCode` value in `build.gradle` for individual extensions
- [No need] Updated `overrideVersionCode` or `baseVersionCode` as needed for all multisrc extensions
- [ ] Referenced all related issues in the PR body (e.g. "Closes #xyz")
- [x] Added the `isNsfw = true` flag in `build.gradle` when appropriate
- [No need] Have not changed source names
- [No need] Have explicitly kept the `id` if a source's name or language were changed
- [x] Have tested the modifications by compiling and running the extension through Android Studio
- [x] Have removed `web_hi_res_512.png` when adding a new extension

SummonHIM committed:
Init commit of the Komiic

SummonHIM committed:
komiic: set ext to nsfw

SummonHIM committed:
Komiic: Add fetchAPILimit

AwkwardPeak7 commented on src/zh/komiic/src/eu/kanade/tachiyomi/extension/zh/komiic/Komiic.kt:
```diff
@@ -0,0 +1,414 @@
+        const val PREFIX_ID_SEARCH = "id:"
+    }
+
+    override val client: OkHttpClient = network.cloudflareClient.newBuilder().build()
```
```suggestion
    override val client: OkHttpClient = network.cloudflareClient
```

AwkwardPeak7 commented on src/zh/komiic/src/eu/kanade/tachiyomi/extension/zh/komiic/Komiic.kt:
```diff
@@ -0,0 +1,414 @@
+class Komiic : HttpSource() {
+    /**
+     * All generate by ChatGPT 4. But, it works...
+     */
```
...

AwkwardPeak7 commented on src/zh/komiic/src/eu/kanade/tachiyomi/extension/zh/komiic/Komiic.kt:
```diff
@@ -0,0 +1,414 @@
+            .headers(headers)
+            .post(Json.encodeToString(JsonObject.serializer(), json).toRequestBody())
+            .build()
+    }
```
use `POST` function instead

AwkwardPeak7 commented on src/zh/komiic/src/eu/kanade/tachiyomi/extension/zh/komiic/Komiic.kt:
```diff
@@ -0,0 +1,414 @@
+                putJsonObject("pagination") { put("limit", 20); put("offset", (page - 1) * 20); put("orderBy", "VIEWS"); put("status", ""); put("asc", true) }
+            }
+            put("query", "query hotComics(\$pagination: Pagination!) {\n  hotComics(pagination: \$pagination) {\n    id\n    title\n    status\n    year\n    imageUrl\n    authors {\n      id\n      name\n      __typename\n    }\n    categories {\n      id\n      name\n      __typename\n    }\n    dateUpdated\n    monthViews\n    views\n    favoriteCount\n    lastBookUpdate\n    lastChapterUpdate ...
+        }
```
should save queries separately and well formatted (see main comment)

AwkwardPeak7 commented on src/zh/komiic/src/eu/kanade/tachiyomi/extension/zh/komiic/Komiic.kt:
```diff
@@ -0,0 +1,414 @@
+            }
+    }
+
+    override fun fetchSearchManga(
```
generally avoid overriding `fetch` methods, and simply use the `request` and corresponding `parse` methods, unless required

AwkwardPeak7 commented on src/zh/komiic/src/eu/kanade/tachiyomi/extension/zh/komiic/Komiic.kt:
```diff
@@ -0,0 +1,414 @@
+        val jsonObj = JSONObject(jsonData).getJSONObject("data").getJSONObject("comicById")
+
+        return SManga.create().apply {
+            url = "https://komiic.com/comic/${jsonObj.getString("id")}"
```
save only the id in db and construct the urls in code (`request` and `getMangaUrl`)

AwkwardPeak7 commented on src/zh/komiic/src/eu/kanade/tachiyomi/extension/zh/komiic/Komiic.kt:
```diff
@@ -0,0 +1,414 @@
+        val sdf = SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss'Z'", Locale.US)
+        sdf.timeZone = TimeZone.getTimeZone("UTC")
+        return sdf.parse(dateStr)?.time ?: 0L
+    }
```
wrap around try catch, read contributing guide for more info

AwkwardPeak7 commented on src/zh/komiic/src/eu/kanade/tachiyomi/extension/zh/komiic/Komiic.kt:
```diff
@@ -0,0 +1,414 @@
+    }
+
+    override fun chapterListParse(response: Response): List<SChapter> {
+        TODO("Not yet implemented")
```
throw `UnsupportedOperationException` instead for unused methods

AwkwardPeak7 reviewed (changes requested):
take a look at [this](https://github.com/keiyoushi/extensions-source/tree/main/src/all/ninenineninehentai/src/eu/kanade/tachiyomi/extension/all/ninenineninehentai) for pointers on graphql and json handling using kotlinx-serialization with serializable classes

SummonHIM committed:
Komiic: Refactor entire project.

AwkwardPeak7 commented on src/zh/komiic/src/eu/kanade/tachiyomi/extension/zh/komiic/Komiic.kt:
```diff
@@ -0,0 +1,291 @@
+        val sdf = SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss'Z'", Locale.US)
+        sdf.timeZone = TimeZone.getTimeZone("UTC")
+        return sdf.parse(dateStr)?.time ?: 0L
+    }
```
save date format as class val and wrap the parsing in try catch

AwkwardPeak7 commented on src/zh/komiic/src/eu/kanade/tachiyomi/extension/zh/komiic/Komiic.kt:
```diff
@@ -0,0 +1,291 @@
+
+        val entries = (tChapters + tBooks).map { chapter ->
+            SChapter.create().apply {
+                url = "$comicUrl/chapter/${chapter.id}/page/1"
```
consider only saving id, otherwise the relative url (also would need to override getChapterUrl for webview url)

SummonHIM replied:
the comicUrl is already relative url
![image](https://github.com/user-attachments/assets/ab4fbd68-88ff-4224-ac0c-7c8f613edb03)

AwkwardPeak7 commented on src/zh/komiic/src/eu/kanade/tachiyomi/extension/zh/komiic/Komiic.kt:
```diff
@@ -0,0 +1,291 @@
+            .map { response ->
+                parseChapterList<ChaptersResponse>(response, manga.url)
+            }
+    }
```
override chapterListRequest method instead

SummonHIM replied:
and api doesn't return the comic id. so i must get this id in manga.url

AwkwardPeak7 replied:
yes, you set the `chapterListRequest` method to return the POST request; then set the `chapterListParse` to parse the response

SummonHIM replied:
chapterListParse's response.request.url only have query url. does't have manga.url

AwkwardPeak7 replied:
oh right didn't see that... you can put it as after a # in apiurl and access it as fragment

AwkwardPeak7 commented on src/zh/komiic/src/eu/kanade/tachiyomi/extension/zh/komiic/Komiic.kt:
```diff
@@ -0,0 +1,291 @@
+        val response = client.newCall(POST(queryAPIUrl, headers, payload)).execute()
+        val limit = response.parseAs<APILimitData>().getImageLimit
+        return limit.limit <= limit.usage
+    }
```
unused?

SummonHIM replied:
like comment. how to throw an exception in reading page? i throw in imageRequest but only have "Failed to load image". how to custom this error text?
![image](https://github.com/user-attachments/assets/53f55553-f7b3-4f69-b87a-76970daf4a29)

AwkwardPeak7 replied:
you can't unfortunately. what is the api limit representing? should it be thrown when page list is fetched?

SummonHIM replied:
> you can't unfortunately. what is the api limit representing? should it be thrown when page list is fetched?

it's comic image request limit.

SummonHIM replied:
> you can't unfortunately. what is the api limit representing? should it be thrown when page list is fetched?

in early commit. it's in the page list fetched. if this limit is reached while reading. reader just output "Failed to load image"

AwkwardPeak7 commented on src/zh/komiic/src/eu/kanade/tachiyomi/extension/zh/komiic/Komiic.kt:
```diff
@@ -0,0 +1,291 @@
+                    )
+                }
+            }
+    }
```
override pageListRequest and pageListParse instead

SummonHIM replied:
this func is same situation as fetchChapterList & parseChapterList

AwkwardPeak7 commented on src/zh/komiic/src/eu/kanade/tachiyomi/extension/zh/komiic/Komiic.kt:
```diff
@@ -0,0 +1,291 @@
+    override fun imageUrlParse(response: Response) = throw UnsupportedOperationException()
+
+    private inline fun <reified T> String.parseAs(): T =
+        Json.decodeFromString(this)
```
```
private val json: Json by injektLazy()
```
put this somewhere in the class, (near the top) and use that instead of directly calling `Json`

AwkwardPeak7 commented on src/zh/komiic/src/eu/kanade/tachiyomi/extension/zh/komiic/Payload.kt:
```diff
@@ -0,0 +1,49 @@
+import kotlinx.serialization.Serializable
+
+@Serializable
+data class Payload<T>(
```
use simple classes, no need for data classes in most cases

SummonHIM replied:
That cause these vars unused warning. is that ok?
![image](https://github.com/user-attachments/assets/42d60b1f-2802-448f-8029-fdd9be5a10fc)

AwkwardPeak7 replied:
Can make them private if they aren't being accessed from outside, and unused warnings on the payload is fine

AwkwardPeak7 commented on src/zh/komiic/src/eu/kanade/tachiyomi/extension/zh/komiic/Response.kt:
```diff
@@ -0,0 +1,168 @@
+
+@Serializable
+data class Comic(
+    @SerialName("id") val id: String,
```
no need for serialname where the property name is same

AwkwardPeak7 commented on src/zh/komiic/src/eu/kanade/tachiyomi/extension/zh/komiic/Response.kt:
```diff
@@ -0,0 +1,168 @@
+
+interface ImagesResult {
+    val images: List<Image>
+}
```
interface is only needed/used for unifying the parsing of popular and latest etc, not need otherwise

see [this](https://github.com/keiyoushi/extensions-source/blob/main/src/all/mangapark) as another example source

SummonHIM committed:
komiic: save date format as class val and wrap the parsing in try catch

SummonHIM committed:
Comiic: remove unnecessary private function
rename some vars
remove unnecessary SerialName
remove unnecessary interface

SummonHIM committed:
komiic:
add private val json
payload use simple classes
change some companion object to capital

SummonHIM committed:
Komiic:
imports go ordered in lexicographic
commet function fetchAPILimit

SummonHIM committed:
Komiic: restore previous import order

AwkwardPeak7 commented on src/zh/komiic/src/eu/kanade/tachiyomi/extension/zh/komiic/Komiic.kt:
```diff
@@ -0,0 +1,285 @@
+    override fun chapterListParse(response: Response): List<SChapter> {
+        val res = response.parseAs<Data<ChaptersResponse>>()
+        val comics = res.data.chapters
+        val comicUrl = response.request.url.toString().split("#")[1]
```
```suggestion
        val comicUrl = response.request.url.fragment
```

AwkwardPeak7 commented on src/zh/komiic/src/eu/kanade/tachiyomi/extension/zh/komiic/Komiic.kt:
```diff
@@ -0,0 +1,285 @@
+            Page(
+                index,
+                "${chapterUrl.substringBeforeLast("/")}/$index",
+                "https://komiic.com/api/image/${image.kid}",
```
```suggestion
                "$baseUrl/api/image/${image.kid}",
```

SummonHIM committed:
Komiic: optimize some variables
